packages/cdk/resources/Cognito.ts

```diff
@@ -0,0 +1,207 @@
+        props.mockOidcClientId === undefined ||
+        props.mockOidClientSecret === undefined ||
+        props.mockOidcIssuer === undefined
+      ) {
+        throw new Error("Attempt to use mock oidc but variables are not defined")
+      }
+
+      // these are the endpoints that are added to user pool identity provider
+      // note we override the token endpoint to point back to our custom token
+      const mockOidcEndpoints: OidcEndpoints = {
+        authorization: props.mockOidcAuthorizeEndpoint,
+        jwksUri: props.mockOidcjwksEndpoint,
+        token: `https://${props.fullCloudfrontDomain}/api/mocktoken`,
+        userInfo: props.mockOidcUserInfoEndpoint
+      }
+
+      mockPoolIdentityProvider = new UserPoolIdentityProviderOidc(this, "MockUserPoolIdentityProvider", {
+        name: "Mock",
+        clientId: props.mockOidcClientId,
+        clientSecret: props.mockOidClientSecret,
+        issuerUrl: props.mockOidcIssuer,
+        userPool: userPool,
+        attributeRequestMethod: OidcAttributeRequestMethod.GET,
+        scopes: ["openid", "profile", "email", "nhsperson", "nationalrbacaccess"],
+        endpoints: mockOidcEndpoints
+      })
+
+      supportedIdentityProviders.push(
+        UserPoolClientIdentityProvider.custom(mockPoolIdentityProvider.providerName)
+      )
+
+    }
+
+    // add attribute mappings to what we receive back to cognito user pool attributes
+    // need to use an escape hatch as can not do this with the L2 construct
+
+    const cfnUserPoolIdentityProvider = primaryPoolIdentityProvider.node.defaultChild as CfnUserPoolIdentityProvider
+    cfnUserPoolIdentityProvider.attributeMapping = {
+      username: "sub",
+      name: "name",
+      given_name: "given_name",
+      family_name: "family_name",
+      email: "email"
+    }
+
+    // add the web client
+    const userPoolWebClient = userPool.addClient("WebClient", {
+      supportedIdentityProviders: supportedIdentityProviders,
+      oAuth: {
+        flows: {
+          authorizationCodeGrant: true,
+          implicitCodeGrant: false
+        },
+        scopes: [
+          OAuthScope.OPENID,
+          OAuthScope.EMAIL,
+          OAuthScope.PHONE,
+          OAuthScope.PROFILE,
+          OAuthScope.COGNITO_ADMIN
+        ],
+        callbackUrls: [
+          "http://localhost:3000/auth/",
+          `https://${props.fullCloudfrontDomain}/site/`,
+          `https://${props.fullCloudfrontDomain}/auth_demo/`
+        ],
+        logoutUrls: [
+          "http://localhost:3000/",
+          `https://${props.fullCloudfrontDomain}/site/`,
+          `https://${props.fullCloudfrontDomain}/auth_demo/`
+        ]
+      }})
+
+    // ensure dependencies are set correctly so items are created in the correct order
+    userPoolWebClient.node.addDependency(primaryPoolIdentityProvider)
+    if (props.useMockOidc) {
+      userPoolWebClient.node.addDependency(mockPoolIdentityProvider)
+    }
```

Comment:
This seems like overkill, they are directly referenced in the definition of the client, do we definitely need the explicit dependency as well?

Reply:
This is one of the times where explicit dependencies are needed to ensure resources are created in the correct order